---
title: "Egger Regression analysis and plots"
author: "Rachael Aikens, Swarthmore College Department of Biology and Mathematics"
date: "June 20, 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##Summary

The following document includes R code to run Egger Regression and Inverse-Variance-weighted (IVW) Regression for MR causal inference and sensitivity analysis.  Additional code is provided to visualize the results in linear regression plots and funnel plots.

##Basic IVW and Egger Regression

As a comparison to Egger Regression, we have provided the code to run IVW regression. This mehtod produces effect estimates asymptotically identical to the GRS analysis.  Begin with input data of exposure (X) and outcome (Y) associations for each SNP in the genetic instrument.  As an example, we will use our SBP and T2D association data for our conservative set of 13 SNPs exclusively associated with SBP.  All of the code from this section is based from the supplimentary code made avalable by Bowden et al. ("Mendelian Randomization with Invalid Instruments", 2015).

```{r, eval=TRUE}
#vector of gene-exposure associations (SBP)
BetaXG <- c(0.357, 0.329, 0.732, 0.903, 0.620, 0.485, 0.355, 0.318, 0.951, 0.553, 0.520, 0.327, 0.471)

#Vector of gene-outcome associations (T2D)
BetaYG <- c(0.0259, -0.0011, 0.0152, 0.0506, 0.0188, 0.0029, -0.0099, -0.0280, 0.0056, 0.0089, 0.0033, -0.0052, 0.0103)

#Vector of standard errors in gene-outcome associations
se.BetaYG <- c(0.0128, 0.0103, 0.0115, 0.0142, 0.0125, 0.0126, 0.0164, 0.0169, 0.0157, 0.0117, 0.0155, 0.0113, 0.0106)
```

Next, we run some pre-processing to ensure that all gene-exposure associations are positive.

```{r}
BYG <- BetaYG*sign(BetaXG) #flip sign where needed in outcome association.
BXG <- abs(BetaXG) #flip to positive in exposure association. 
```

Run IVW regression. These calculations use a quick asymptotic correction for estimating standard error.

```{r}
#IVW
ivwlm <- lm(BetaYG~ -1 + BetaXG,weights=1/se.BetaYG^2)
IVWfit <- summary(ivwlm)
IVWBeta <- IVWfit$coef[1,1]

#asymptotic correction for SE
SE <- IVWfit$coef[1,2]/IVWfit$sigma

#p values and CIs from t distribution
DF <- length(BetaYG)-1
IVW_p = 2*(1-pt(abs(IVWBeta/SE),DF))
IVW_CI = IVWBeta + c(-1,1)*qt(df=DF,0.975)*SE

#report results
IVWResults = c(IVWBeta,SE,IVW_CI,IVWBeta/SE,IVW_p)
names(IVWResults)=c("Point Estimate","Corrected Standard Error", "95% CI: lower", "95% CI: upper", "t-statistic","p value")

IVWResults
```
Run Egger regression, also using the asymptotic standard error correction.

```{r}
#Egger regression
erlm <- lm(BYG~BXG,weights=1/se.BetaYG^2)
ERfit <- summary(erlm)
EggerBeta0 <- ERfit$coef[1,1]
EggerBeta1 <- ERfit$coef[2,1]

#asymptotic correction for SE
SE0 <- ERfit$coef[1,2]/ERfit$sigma
SE1 <- ERfit$coef[2,2]/ERfit$sigma

#p values and CIs from t distribution
DF <- length(BetaYG)-2
ERBeta0_p <- 2*(1-pt(abs(EggerBeta0/SE0),DF))
ERBeta1_p <- 2*(1-pt(abs(EggerBeta1/SE1),DF))
ERBeta0_CI <- EggerBeta0 + c(-1,1)*qt(df=DF, 0.975)*SE0
ERBeta1_CI <- EggerBeta1 + c(-1,1)*qt(df=DF, 0.975)*SE1

#report results
EggerResults <- matrix(nrow=2, ncol = 6)
EggerResults[1,] <- c(EggerBeta0, SE0, ERBeta0_CI, EggerBeta0/SE0,ERBeta0_p)
EggerResults[2,] <- c(EggerBeta1, SE1, ERBeta1_CI, EggerBeta1/SE1,ERBeta1_p)
rownames(EggerResults) <- c("Intercept(Bias)", "Effect Estimate")
colnames(EggerResults) <- c("Point Estimate","Corrected Standard Error","95% CI: lower", "95% CI: upper", "t-statistic","p value")

EggerResults
```

##Visualization with Regression and Funnel Plots

Funnel plots are a common tool in MR to quickly visualize SNP associations and detect outliers.  The code below generates a funnel plot with vertical lines displaying the exposure-outcome effect estimates generated by IVW and Effer regression.  For these plots, we use Gene-exposure values which are corrected based on effect allele frequency:

```{r, eval = TRUE}
#vector of effect allele frequencies from sample data
eaf <- c(0.3298, 0.4657, 0.2770, 0.1372, 0.3232, 0.2968, 0.4802, 0.3670, 0.8602, 0.7177, 0.1425, 0.6214, 0.4551)
```
In accordance with Bowden et al. 2015, a SNP with estimated gene-exposure association $\hat{\gamma}_j$ and estimated effect allele frequency $\hat{q}_j$ has corrected association $$\hat{\gamma}^C_j = \hat{\gamma}_j\sqrt{\hat{q}_j(1-\hat{q}_j)}$$

```{r, eval = TRUE}
BetaXG.C <- abs(BetaXG*sqrt(eaf*(1-eaf)))
```
Use ggplot2 to generate funnel plot (The ggplot2 package must be installed to run this code):

```{r, eval = TRUE}
library(ggplot2)

```
```{r, eval = TRUE}
#Funnel plot
BetaXY=BYG/BXG
funnel <- qplot(BetaXY,BetaXG.C,size=I(2.5)) +
  labs(x=expression("Effect Estimate (" ~ hat(beta)["j"] ~")"), y=expression("Gene-exposure association (" ~ hat(gamma)["j"]^"C" ~ ")"))+
  geom_vline(aes(xintercept = IVWBeta, color = "IVW"), show.legend = TRUE)+
  geom_vline(aes(xintercept=EggerBeta1,color = "Egger Regression"), show.legend = TRUE)+
  scale_color_manual("",values=c("blue","red"))+
  theme(legend.title = element_blank(), 
        legend.justification= c(0,1), 
        legend.position=c(0,1),
        legend.text = element_text(size=14),
        axis.text=element_text(size=12),
        text = element_text(size=17))

funnel
```

MR regression plots can be used to visualize the estimated effect and bias.  The slopes of the IVW and Egger regression lines correspond to their effect estimates.  The y-intercept of the Egger regression line serves as an estimate of bias due to pleiotropy.

```{r, eval = TRUE}
#Regression plot
mrplot = qplot(BXG,BYG,size=I(2.5), xlim=c(0,1.2)) +
  labs(x=expression("Association with SBP (" ~ hat(gamma)["j"] ~")"), y=expression("Association with T2D Risk (" ~ hat(Gamma)["j"] ~")"))+
  geom_abline(aes(intercept=coef(erlm)[[1]],slope=coef(erlm)[[2]], color = "Egger Regression"),show.legend = FALSE) +
  geom_abline(aes(intercept=0,slope=coef(ivwlm)[[1]], color = "IVW"),show.legend=TRUE) +
  scale_color_manual("",values=c("blue","red"))+
  theme(legend.title = element_blank(), 
        legend.justification= c(1,0), 
        legend.position=c(1,0),
        legend.text = element_text(size = 14),
        axis.text=element_text(size=10),
        text = element_text(size = 17))+
  geom_hline(yintercept = 0, color ="black") +
  geom_vline(xintercept = 0, color = "black")

mrplot
```

##Bootstrap Corrections for Regression Standard Error

The standard errors for IVW and Egger Regression can also be ascertained by bootstrapping.  The code below calculates bootstrap estimates of standard error and 95% confidence intervals.  First, we need to know the standard errors in gene-exposure estimates: 

```{r, eval = TRUE}
se.BetaXG <- c(0.079, 0.077, 0.079, 0.179, 0.086, 0.097, 0.068, 0.069, 0.113, 0.0838, 0.108, 0.067, 0.076)
```
Since the remainder of this analysis is more computationally intensive, these r chunks have been set so that they are not evaluated.

```{r, eval = FALSE}
#Bootstrap correction for IVW
bootivw <- NULL; straps = 10000
for (i in 1:straps) {
  BYG_boot = rnorm(length(BYG), mean 
                   =BYG, sd=se.BetaYG)
  BXG_boot = rnorm(n = length(BXG), mean=BXG, sd=se.BetaXG)
  BYG_boot = BYG_boot*sign(BXG_boot)
  BXG_boot = abs(BXG_boot)
  bootivw[i] = summary(lm(BYG_boot~ -1 + BXG_boot,weights=se.BetaYG^-2))$coef[1,1]
}
boot_upperivw = sort(bootivw)[9751]
boot_lowerivw = sort(bootivw)[250]
boot_seivw = sd(bootivw)
IVWResultsBoot = c(IVWBeta,boot_seivw,boot_lowerivw,boot_upperivw)
names(IVWResultsBoot)=c("Point Estimate","Corrected Standard Error", "95% CI: lower", "95% CI: upper")

IVWResultsBoot

#bootstrap correction for Egger Regression: Beta0(intercept)
boot0 <- NULL; straps = 10000
for (i in 1:straps) {
  BYG_boot = rnorm(length(BYG), mean 
                   =BYG, sd=se.BetaYG)
  BXG_boot = rnorm(length(BXG), mean =BXG, sd=se.BetaXG)
  BYG_boot = BYG_boot*sign(BXG_boot)
  BXG_boot = abs(BXG_boot)
  boot0[i] = summary(lm(BYG_boot~BXG_boot,weights=se.BetaYG^-2))$coef[1,1]
}
boot_upper0 = sort(boot0)[9751]
boot_lower0 = sort(boot0)[250]
boot_se0 = sd(boot0)

#bootstrap correction for Egger Regression: Beta1(slope)
boot1 <- NULL; straps = 10000
for (i in 1:straps) {
  BYG_boot = rnorm(length(BYG), mean =BYG, sd=se.BetaYG)
  BXG_boot = rnorm(length(BXG), mean = BXG, sd = se.BetaXG)
  BYG_boot = BYG_boot*sign(BXG_boot)
  BXG_boot = abs(BXG_boot)
  boot1[i] = summary(lm(BYG_boot~BXG_boot,weights=se.BetaYG^-2))$coef[2,1]
}
boot_upper1 = sort(boot1)[9751]
boot_lower1 = sort(boot1)[250]
boot_se1 = sd(boot1)



#results for boot correction into matrix
EggerBoot <- matrix(nrow=2, ncol = 4)
EggerBoot[1,] <- c(EggerBeta0, boot_se0, boot_lower0, boot_upper0)
EggerBoot[2,] <- c(EggerBeta1, boot_se1, boot_lower1, boot_upper1)
rownames(EggerBoot) <- c("Intercept(Bias)", "Effect Estimate")
colnames(EggerBoot) <- c("Point Estimate","Boot Corrected Standard Error","95% CI: lower", "95% CI: upper")

EggerBoot
```